### DISCUSSION TITLE
Pointers to what works well with GPT function/tool calling?

### DISCUSSION THREAD

**Discussion by indigoviolet** [567-labs/instructor#332] [Q&A] [answered]
I have seen the [Tips](https://jxnl.github.io/instructor/concepts/prompting/) page.

Have you come across pointers around what function calling patterns are most effective (my interest currently is with OpenAI models)? 

Especially with regards to understanding how the JSON representation of the function/tool is communicated to the model, and if there is any signal loss in that process? Is JSON Schema actually supported or only a subset? Which types work well, how much nesting, which titles/descriptions/docstrings etc?

**Comment by jxnl** [comment, marked as answer]
Docstrings are very important, but everything I know I've written down in the documentation.

**Comment by jxnl** [comment]
Tools and function calls are compiled to a type of representation, so there is some loss
